packages/firebase_messaging/android/src/main/java/io/flutter/plugins/firebasemessaging/FlutterFirebaseMessagingService.java

```diff
@@ -18,16 +39,81 @@
   public static final String ACTION_TOKEN = "io.flutter.plugins.firebasemessaging.TOKEN";
   public static final String EXTRA_TOKEN = "token";
 
+  private static final String SHARED_PREFERENCES_KEY = "io.flutter.android_fcm_plugin";
+  private static final String BACKGROUND_SETUP_CALLBACK_HANDLE_KEY = "background_setup_callback";
+  private static final String BACKGROUND_MESSAGE_CALLBACK_HANDLE_KEY =
+      "background_message_callback";
+
+  // TODO(kroikie): make sIsIsolateRunning per-instance, not static.
+  private static AtomicBoolean sIsIsolateRunning = new AtomicBoolean(false);
+
+  /** Background Dart execution context. */
+  private static FlutterNativeView sBackgroundFlutterView;
```

Comment:
@collinjackson @kroikie I'm guessing that most of this stuff was copied from android_alarm_manager, right? If so, what do you two think about extracting the commonalities? There are enough regrettable details in the android_alarm_manager implementation that it might be a good idea to find a way to centralize it so that we can refactor and improve it one time. I'm not sure how we'd do it, or where the correct encapsulation boundary is, but I thought it worth asking.

Reply:
I would support having the common functionality extracted. I'm not sure if now's the time to do it, or if we should land this functionality and with a TODO to refactor plugins as a second pass. Are there other important use cases for a background Dart isolate other than handling local and remote messages? If it's just these two plugins, then perhaps putting a lot of effort into building a common API is less critical.

Reply:
All mobile app background behavior will eventually become a candidate for this kind of implementation...or preferably a much better one. This could be any OS signal: push notifications, user location changes, bluetooth messages, possibly 2FA messages, etc. Any message that you can register with the OS to receive could theoretically require background processing and full-Flutter apps would probably want to execute the application logic in Dart.

I've actually been suggesting since late last year that we consider background processing in Android to be its own project, fully researching the issue, finding relevant customers, and then producing a strong architectural solution. Obviously we can't do that here, but that's the background context (pun intended).

Reply:
It sounds like a worthwhile project, the question in my mind is whether it should block landing this PR. Are you leaning one way or another?

Reply:
We definitely don't need to hold up this PR on that much larger project. But we should all be aware that this behavior, both here and in android_alarm_manager, will likely need major refactoring once we have a better solution. 

As for centralizing this behavior, I'll leave that judgement to you and @kroikie because you two have better visibility into the desired timeline for this work. If we choose not to centralize this background behavior, then we should all be sure that any fix or change to any of this behavior in one project should also be made in the other project to avoid divergence.

Reply:
OK, sounds like we should move forward with this PR with a TODO and file an issue for centralizing the behavior.

Reply:
Agreed!